### PR TITLE
InfoComponentVisitor makes DetectorInfo and ComponentInfo

### LINKED ISSUE #20002: Move DetectorInfo into InfoComponentVisitor
This is a grouping or related items taken from the umbrella Item listed here: #19964

* `InfoComponentVisitor` to create `DetectorInfo` as well as `ComponentInfo` (1 from umbrella list)
* `Beamline::DetectorInfo` and `Beamline::ComponentInfo` should be created together. Construction should be as safe as possible as both have a runtime dependency on each other (2 from umbrella list)
* Remove work-around code in ComponentInfo::getPos etc, that tries to deal with current staged creation of the `..Info` objects. (7 from umbrella list)
* Possibly link these items via some other structure to ensure that lifetime rules (internal pointer ownership of each) make better sense. This is is a design aspect that falls out of the above. I am proposing a grouping called `Beamline` (agreed name for Instrument 2.0), that holds and provides well established semantics for `ComponentInfo` and `DetectorInfo`.

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #20002. Make `ComponentInfo` and `DetectorInfo` via visitor double dispatch mechanism.

WIP

Check key performance areas such as `CalibrateSXDPanels` and common Load/Save algorithms. Should be at least as good as base branch `instrument_2_0_1` 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Beamline/inc/MantidBeamline/Beamline.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class MANTID_BEAMLINE_DLL Beamline {
```
Comment (SimonHeybrock) [inline Framework/Beamline/inc/MantidBeamline/Beamline.h]: I think we will be running into trouble with this name -- there will be clashes with the namespace of the same name (I had this issue with `Histogram`, this is why the module is names `HistogramData`.

File: Framework/Beamline/inc/MantidBeamline/Beamline.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  Beamline alias();
+  bool empty() const;
+
+private:
+  // For internal use only.
+  Beamline(boost::shared_ptr<ComponentInfo> &componentInfo,
+           boost::shared_ptr<DetectorInfo> &detectorInfo);
+  bool m_empty;
+  boost::shared_ptr<ComponentInfo> m_componentInfo;
+  boost::shared_ptr<DetectorInfo> m_detectorInfo;
```
Comment (SimonHeybrock) [inline Framework/Beamline/inc/MantidBeamline/Beamline.h]: Why do we need to store these as `shared_ptr`? The data inside `ComponentInfo` and `DetectorInfo` is shared, so it is not clear why we would want another level of sharing.

File: Framework/Geometry/inc/MantidGeometry/Instrument.h [hunk 3]
```diff
@@ -6,25 +6,22 @@
 #include "MantidGeometry/IDetector.h"
 #include "MantidGeometry/Instrument/CompAssembly.h"
 #include "MantidGeometry/Instrument/ObjComponent.h"
-
 #include "MantidKernel/DateAndTime.h"
+#include "MantidBeamline/Beamline.h"
```
Comment (SimonHeybrock) [inline Framework/Geometry/inc/MantidGeometry/Instrument.h]: Consider avoiding this include and instead use a forward declaration (and store `Beamline` as `unique_ptr`).

File: Framework/Geometry/inc/MantidGeometry/Instrument/ParameterMap.h [hunk 4]
```diff
@@ -6,6 +6,7 @@
 #include "MantidGeometry/IDetector.h"
 #include "MantidGeometry/IDTypes.h" //For specnum_t
 #include "MantidGeometry/Instrument/Parameter.h"
+#include "MantidBeamline/Beamline.h"
```
Comment (SimonHeybrock) [inline Framework/Geometry/inc/MantidGeometry/Instrument/ParameterMap.h]: Maybe also avoid this include?

File: Framework/Geometry/src/Instrument.cpp [hunk 5]
```diff
@@ -1317,14 +1286,15 @@ boost::shared_ptr<ParameterMap> Instrument::makeLegacyParameterMap() const {
...
-  pmap->setComponentInfo(nullptr);
+  pmap->clearBeamline();
   // Instrument is only needed for DetectorInfo access so it is not needed.
   pmap->setInstrument(nullptr);
 
-  if (!hasDetectorInfo() && !hasComponentInfo())
+  if (!hasInstrumentVisitor())
     return pmap;
 
+  auto componentCache = m_instrVisitor->componentIds();
```
Comment (SimonHeybrock) [inline Framework/Geometry/src/Instrument.cpp]: `auto &`?

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Looks pretty good already. My main comment is regarding the copy/ownership mechanism in `Beamline::Beamline` that you already mentioned on Skype. I do not really understand the mechanism. It seems to be made for handling cases where one of `ComponentInfo` and `DetectorInfo` can be `nullptr`? What is that used for?

File: Framework/Beamline/inc/MantidBeamline/Beamline.h (hunk 1 above)
Comment (OwenArnold) [inline Framework/Beamline/inc/MantidBeamline/Beamline.h]: There was a previous chat on the general slack thread, and I think Duc suggested "Beamline" a name. Instrument 2.0 doesn't seem to cut, it and `Instrument` is too easily confused with existing `Instrument 1.0` IMHO. I do agree that `Beamline` itself might not be descriptive enough. Open for suggestions... maybe

* `Geometry`  - as in `Beamline::Geometry`, too easily confused with namespace `Geometry`?
* `BeamlineInstrument`?

File: Framework/Beamline/inc/MantidBeamline/Beamline.h (hunk 2 above)
Comment (OwenArnold) [inline Framework/Beamline/inc/MantidBeamline/Beamline.h]: I had to change it from a `unique_ptr` to this. I think i'm going to write down the various object/ownership issues I encountered. I did add a regression tests for each of these as I came across and fixed these. I think some of your questions will relate to these so I'll write it properly once.

File: Framework/Geometry/inc/MantidGeometry/Instrument.h (hunk 3 above)
Comment (OwenArnold) [inline Framework/Geometry/inc/MantidGeometry/Instrument.h]: Considered that, and sat of the fence for a while - eventually decided that ownership and copying semantics issues were complex enough that I wanted to shield from the client. will define further below.

File: Framework/Beamline/inc/MantidBeamline/Beamline.h (hunk 1 above)
Comment (SimonHeybrock) [inline Framework/Beamline/inc/MantidBeamline/Beamline.h]: The discussion in slack was when I looked for a module name, and indeed I then used Duc's suggestion `Beamline`.

- `Geometry` has nothing to do with neutron instruments.
- `BeamlineInstrument` does not make a lot of sense?
- `Instrument` does not seem too bad for me. The existing `Geometry::Instrument` is being removed after all.

Personally I would, I think, also prefer `Beamline::Beamline`, but I think that implies that we can never use `using namespace Beamline`? Is that an issue?

File: Framework/Geometry/src/Instrument.cpp (hunk 5 above)
Comment (OwenArnold) [inline Framework/Geometry/src/Instrument.cpp]: 👍

**Comment by OwenArnold** [comment]
1. **Deep Copy**

Ensuring correct COW of held requires invoking the copy constructor of DetectorInfo and ComponentInfo. For example, when an `ExperimentInfo` is copied, if the COW ptr holding the vector of positions is not reference count incremented, then modifying positions of the first `Workspace`/`ExperiementInfo` will result in positions on the second incorrectly being changed.

```cpp
ExperimentInfo a;
...
ExperimentInfo b = a; // Deep copy in beamline required here!
a.beamline().detectorInfo().setPosition(V3D{1,1,1}); // Otherwise sharing in b not broken.
```

Current solution: Copy constructors of held pointers (`DetectorInfo` and `ComponentInfo`) in `Beamline` must be invoked on a `Beamline` copy

2. **Shallow Copy**

The original design had `ExperimentInfo` as the owner of `Beamline`. References loaned out to `ParameterMap` as part of `ExperimentInfo::setInstrument`. However, th there are cases where the Instrument is used after the ExperimentInfo has goine out of scope. While we do NOT want a deep copy, we need to ensure that the ParameterMap can have shared ownership of the beamline. See https://github.com/mantidproject/mantid/blob/8c85f9a168ef352e6dc9cbebdb03cc58fee69b15/Framework/API/test/ExperimentInfoTest.h#L974-L999

Current solution: `Beamline` provides an `alias` member function which calls a private constructor, which copies the held pointers but elides their copy constructors. Ownership of member pointers is shared so that use cases such as: `ParameterMap` can remain valid after creating `ExperimentInfo` goes out of scope.

3. **Move construction and assignment**

For convenience we might want to move an existing named beamline object, or reuse a temporary. 

Current solution: A default move constructor and move assignment operator are declared on `Beamline`.

**Comment by SimonHeybrock** [comment]
I think this has been dealt with in another PR, closing?

**Pull request closed without merging**